### PR TITLE
feature #118: Create an admin groups flow

### LINKED ISSUE #118: [FEATURE] Create an admin groups flow
We need to create "Search groups", "Edit group", "Create group" pages

## Front-end part

Search groups:

- Sidebar that is already implemented
- Search panel:
  - Search input that searches by code of group
  - Filter by year of group (multichoice dropdown with search)
  - Filter by specialty of group (multichoice dropdown with search)
  - Filter by cathedra of group (multichoice dropdown with search)
  - Sorting input with variants "by code", "by year", "by captain full name"
  - Icon button to change the direction of sorting (asc/desc)
- Button "Create" which opens "Create grant" page
- Table:
  - group code (string)
  - year (int from 1 to 4)
  - specialty (int)
  - cathedra (in form of tag, abbreaviation), add tooltip with full name and division
  - Button "Edit" which opens "Edit group" page
  - Button "Delete" which opens confirmation pop-up
  - Below the table there are settings for page size and page (from MUI component)
- Back-end will NOT send year of group (number from 1 to 4). It sends year of admission (2021, 2020, 2023, 2022 etc.) You need to get current education year and then find the difference that you will show :>

image

Edit group:

- Sidebar that is already implemented
- Button "Cancel" which opens "Search group" page
- Button "Delete" which opens confirmation pop-up. After deleting opens "Search group" page
- Button "Save" which send a request on back-end endpoint which saves data. After saving opens "Search group" page
- Input for group name (min 1, max 10)
- Dropdown for year of admission (from 2018 to 2025)
- Dropdown for educational program
- Dropdown for cathedra
- Dropdown for captain of group 
- Dropdowns for moderators of group 
- Button "Add moderator" if user want to add another one moderator

image

## Back-end part

- Endpoint for getting groups:
  - filters:
    - years array (int values from 1 to 4)
    - specialties array (ids)
    - cathedra array (ids)
    - code string
  - sorts:
    - code
    - year of admission
    - captain last name
  - must return:
    - id
    - code
    - captain user (id, last, first, middle name, etc.)
    - cathedra (id, name, abbreviation, division)
    - year of admission
    - specialty

- Endpoint for getting cathedras

- Endpoint for getting educational programs

- Endpoint for getting specialties

- Endpoint for editing group (by id):
  - code (min 1, max 10)
  - year of admission
  - educational program (id)
  - captain (by student id) (must be user of this group). Old captain must became student
  - moderator array (ids) (must be users of this group). the logic is that you need to set these moderators and remove another one
  - cathedra (id)

- Endpoint for creating group (by id):
  - code (min 1, max 10)
  - year of admission
  - educational program (id)
  - cathedra (id)

- Endpoint for getting group members with their roles

- Endpoint for deleting group (by id):
  - student must become with null group id and state DECLINED 
  - it's a very risky endpoint. i don't really know do we need to implement it. really risky.

### PR DESCRIPTION
**Pull request by ArtemLyah**
- added admissionYear field in students and groups models
- set groups field optional in student model
- provided migration
- set right admissionYear in database
- updated endpoint for getting all groups
- updated endpoint for creating group
- updated endpoint for updating group
- updated endpoint for deleting group
- created controller for educational program
- updated documentation

closes #118

### CODE REVIEW

File: fictadvisor-api/src/v2/api/controllers/EduProgramController.ts [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+    return this.eduProgramMapper.getPrograms(programs);
+  }
+
+  @ApiOkResponse({
+    type: SpecialityRepository,
+  })
+  @ApiEndpoint({
+    summary: 'Get all specialities',
+  })
+  @Get('/specialitites')
```
Comment (fokaaas) [inline fictadvisor-api/src/v2/api/controllers/EduProgramController.ts:39]: it doesn't belong to this controller

File: fictadvisor-api/src/v2/api/controllers/EduProgramController.ts [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+import { EduProgramMapper } from '../../mappers/EduProgramMapper';
+import { ApiEndpoint } from '../../utils/documentation/decorators';
+import { EduProgramResponses } from '../responses/EduProgramResponse';
+import { SpecialityResponses } from '../responses/SpecialityResponse';
+import { EduProgramService } from '../services/EduProgramService';
+
+@ApiTags('eduprogram')
+@Controller({
+  version: '2',
+  path: '/eduProgram',
```
Comment (fokaaas) [inline fictadvisor-api/src/v2/api/controllers/EduProgramController.ts:13]: `eduprograms`

File: fictadvisor-api/src/v2/api/controllers/EduProgramController.ts [hunk 3]
```diff
@@ -0,0 +1,44 @@
+import { Controller, Get } from '@nestjs/common';
+import { ApiOkResponse, ApiTags } from '@nestjs/swagger';
+import { SpecialityRepository } from '../../database/repositories/SpecialityRepository';
+import { EduProgramMapper } from '../../mappers/EduProgramMapper';
+import { ApiEndpoint } from '../../utils/documentation/decorators';
+import { EduProgramResponses } from '../responses/EduProgramResponse';
+import { SpecialityResponses } from '../responses/SpecialityResponse';
+import { EduProgramService } from '../services/EduProgramService';
+
+@ApiTags('eduprogram')
```
Comment (fokaaas) [inline fictadvisor-api/src/v2/api/controllers/EduProgramController.ts:10]: Eduprogram

File: fictadvisor-api/src/v2/api/controllers/GroupController.ts [hunk 4]
```diff
@@ -50,13 +53,17 @@ export class GroupController {
...
-    type: GroupResponse,
+    type: MappedGroupResponse,
   })
   @ApiBadRequestResponse({
     description: `\n
     InvalidBodyException: 
       Proper name is expected
-      Code can not be empty`,
+      Code can not be empty
+      Educatioal program id cannot be empty
```
Comment (fokaaas) [inline fictadvisor-api/src/v2/api/controllers/GroupController.ts:63]: Educational

File: fictadvisor-api/src/v2/api/dtos/QueryAllGroupDTO.ts [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+    description: 'Array of cources',
+    type: [Number],
+  })
+  @Type(() => Number)
+  @IsArray(validationOptionsMsg('Cources must be an array'))
+  @IsNumber({}, validationOptionsMsg('Each cource must be a number', true))
+  @Min(1, validationOptionsMsg('Min course value is 1', true))
+  @Max(4, validationOptionsMsg('Max course value is 4', true))
+  @IsOptional()
+    cources?: number[];
```
Comment (fokaaas) [inline fictadvisor-api/src/v2/api/dtos/QueryAllGroupDTO.ts:40]: courses

File: fictadvisor-api/src/v2/api/responses/SpecialityResponse.ts [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+    description: 'Speciality name',
+  })
+    name: string;
+} 
+
+export class SpecialityResponses {
+  @ApiProperty({
+    description: 'Array of specialities',
+  })
+  specialities: SpecialityResponse[];
```
Comment (fokaaas) [inline fictadvisor-api/src/v2/api/responses/SpecialityResponse.ts:31]: add indent

File: fictadvisor-api/src/v2/api/services/EduProgramService.ts [hunk 7]
```diff
@@ -0,0 +1,19 @@
...
+    private readonly specialityRepository: SpecialityRepository,
+  ) {}
+
+  getPrograms () {
+    return this.eduProgramRepository.findMany();
+  }
+
+  getSpecialities () {
+    return this.specialityRepository.findMany();
+  }
```
Comment (fokaaas) [inline fictadvisor-api/src/v2/api/services/EduProgramService.ts:18]: that method doesn't belong to this service

File: fictadvisor-api/src/v2/database/entities/DbGroup.ts [hunk 8]
```diff
@@ -1,9 +1,21 @@
...
   telegramGroups: TelegramGroup[];
+  cathedra?: Cathedra;
+  educationalProgram?: {
+    speciality: Speciality,
+  } & EducationalPrograms
+  students: ({
+    roles?: {
+      studentId: string,
+      roleId: string,
+      role: Role,
```
Comment (fokaaas) [inline fictadvisor-api/src/v2/database/entities/DbGroup.ts:17]:
```ts
roles?: ({
      role: Role,
    } & UserRole)[]
```

**Review by fokaaas — CHANGES_REQUESTED** [review]
Configure the linter!

**Review by fokaaas — CHANGES_REQUESTED** [review]

File: fictadvisor-api/src/v2/api/services/GroupService.ts [hunk 9]
```diff
@@ -71,23 +74,104 @@ export class GroupService {
...
+    cathedras: (cathedras: string[]) => ({
+      cathedra: {
+        id: {
+          in: cathedras,
+        },
+      },
+    }),
+    cources: (cources: number[]) => {
+      const courceDate = new Date();
+      courceDate.setMonth(courceDate.getMonth()+4);
```
Comment (fokaaas) [inline fictadvisor-api/src/v2/api/services/GroupService.ts:158]: magic number

**Review by fokaaas — CHANGES_REQUESTED** [review]

**Review by igaryakqwe — DISMISSED** [review]

**Review by igaryakqwe — APPROVED** [review]

**Review by f-ogoog — APPROVED** [review]

**Review by igaryakqwe — APPROVED** [review]

**Review by f-ogoog — APPROVED** [review]

**Pull request merged**